### ISSUE TITLE
fix(converters): Converters are called twice

### ISSUE THREAD

**Issue by davinkevin** [jOOQ/jOOQ#10464]
### Expected behavior and actual behavior:

When using converters, I expect them to be called only once per "usage". 

In my use case, I use converters to encrypt/decrypt data before calling SQL with JOOQ. This removes all complexity around encryption/decryption from other code layers. 

The problem is the converter code can take some time to execute and JOOQ seems to call it twice in a query.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

1. Create a converter with a log in the `to` method
2. Declare it in the JOOQ generator or use it with the `DSL.value` method
3. Execute a simple query which use the converter on a field in the `where` predicate

For example: 

```java
query
                .select(FOO.A, FOO.B)
                .from(FOO)
                .where(FOO.ENCRYPTED_FIELD.eq(clearValue))
                .limit(1)
                .toMono()
```

Here, the `FOO.ENCRYPTED_FIELD` use the declared converted created in step (1). It is generated as this in the `Foo` class:

```java
public final TableField<FooRecord, String> ENCRYPTED_FIELD = createField(DSL.name("Foo"), org.jooq.impl.SQLDataType.VARCHAR(128).nullable(false), this, "", new EncryptionConverter());
```

FYI, no other fields in the class use this converter. 

If I read the log of the application, I have this: 

```log
2020-08-03 12:12:46.636  INFO 72857 --- [oundedElastic-1] c.a.i.a.c.d.c.EncryptionConverter  : before encryption VALUE
2020-08-03 12:12:47.541  INFO 72857 --- [oundedElastic-1] c.a.i.a.c.d.c.EncryptionConverter  : after encryption
2020-08-03 12:12:47.552  INFO 72857 --- [oundedElastic-1] c.a.i.a.c.d.c.EncryptionConverter  : before encryption VALUE
2020-08-03 12:12:48.249  INFO 72857 --- [oundedElastic-1] c.a.i.a.c.d.c.EncryptionConverter  : after encryption
```

### Versions:

- jOOQ: 3.13.2
- Java: 11.0.5
- Database (include vendor): PostgreSQL 9.6
- OS: MacOS, Linux…
- JDBC Driver (include name if inofficial driver): pg jdbc 42.2.12

**Comment by lukaseder** [comment]
Thank you very much for your report.

There should be only one call per usage as you're expecting. However, by default, jOOQ enables execute logging: https://www.jooq.org/doc/latest/manual/sql-building/dsl-context/custom-settings/settings-execute-logging

If you're logging on the `DEBUG` level, the SQL query is generated twice: 1) for the prepared statement (with bind values), 2) for the log output (with inlined bind values).

Can you confirm this is the case? If not, what does your `toMono()` call do?

**Comment by lukaseder** [comment]
My bad. While logging can cause converters to be called multiple times as well, they're called twice also because bind values are traversed twice:

1. To generate the SQL string
2. To send the bind value to the JDBC driver

A quick test:

```java
public void testConvertOnlyOnce() throws Exception {

    // [#10464] Make sure Converter::from and ::to are called only once
    bindAndInline(ctx -> {
        ctx.settings().setExecuteLogging(false);
        ctx.configuration().set(new ExecuteListener[0]);

        int[] from = new int[1];
        int[] to = new int[1];

        DataType<String> type = VARCHAR.asConvertedDataType(Converter.ofNullable(
            String.class,
            String.class,
            s -> {
                new Exception().printStackTrace();
                from[0]++;
                return "b";
            },
            s -> {
                new Exception().printStackTrace();
                to[0]++;
                return "c";
            }
        ));

        assertEquals("b", ctx.fetchValue(val("a", type)));
        assertEquals(1, from[0]);
        assertEquals(1, to[0]);
    });
}
```

Fails as `to()` is called twice, and the stack traces are:

```
java.lang.Exception
	at org.jooq.test.all.testcases.ConverterTests.lambda$7(ConverterTests.java:256)
	at org.jooq.Converter.lambda$3(Converter.java:240)
	at org.jooq.Converter$1.to(Converter.java:173)
	at org.jooq.impl.DefaultBinding$AbstractBinding.sql(DefaultBinding.java:780)
	at org.jooq.impl.Val.accept(Val.java:130)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:609)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:226)
	at org.jooq.impl.QueryPartCollectionView.accept(QueryPartCollectionView.java:164)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:609)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:264)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:1587)
	at org.jooq.impl.SelectQueryImpl.toSQLReference0(SelectQueryImpl.java:1426)
	at org.jooq.impl.SelectQueryImpl.toSQLReferenceLimitDefault(SelectQueryImpl.java:1226)
	at org.jooq.impl.SelectQueryImpl.accept0(SelectQueryImpl.java:1128)
	at org.jooq.impl.SelectQueryImpl.accept(SelectQueryImpl.java:914)
	at org.jooq.impl.DefaultRenderContext.visit0(DefaultRenderContext.java:609)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:264)
	at org.jooq.impl.AbstractQuery.getSQL0(AbstractQuery.java:524)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:326)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:481)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:450)
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:464)
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:641)
	at org.jooq.impl.SelectImpl.fetchOne(SelectImpl.java:3074)
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4535)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4573)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4587)
	at org.jooq.test.all.testcases.ConverterTests.lambda$5(ConverterTests.java:262)
	at org.jooq.test.jOOQAbstractTest.bindAndInline(jOOQAbstractTest.java:1127)
	at org.jooq.test.BaseTest.bindAndInline(BaseTest.java:1628)
	at org.jooq.test.all.testcases.ConverterTests.testConvertOnlyOnce(ConverterTests.java:240)
	at org.jooq.test.jOOQAbstractTest.testConvertOnlyOnce(jOOQAbstractTest.java:5422)
```

And then:

```
java.lang.Exception
	at org.jooq.test.all.testcases.ConverterTests.lambda$7(ConverterTests.java:256)
	at org.jooq.Converter.lambda$3(Converter.java:240)
	at org.jooq.Converter$1.to(Converter.java:173)
	at org.jooq.impl.DefaultBinding$AbstractBinding.set(DefaultBinding.java:835)
	at org.jooq.impl.DefaultBindContext.bindValue0(DefaultBindContext.java:62)
	at org.jooq.impl.AbstractBindContext.bindValue(AbstractBindContext.java:124)
	at org.jooq.impl.Val.accept(Val.java:143)
	at org.jooq.impl.AbstractBindContext.bindInternal(AbstractBindContext.java:279)
	at org.jooq.impl.AbstractBindContext.visit0(AbstractBindContext.java:88)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:264)
	at org.jooq.impl.QueryPartCollectionView.accept(QueryPartCollectionView.java:164)
	at org.jooq.impl.AbstractBindContext.bindInternal(AbstractBindContext.java:279)
	at org.jooq.impl.AbstractBindContext.visit0(AbstractBindContext.java:88)
	at org.jooq.impl.AbstractContext.visit(AbstractContext.java:264)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:367)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:481)
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:450)
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:464)
	at org.jooq.impl.AbstractResultQuery.fetchOne(AbstractResultQuery.java:641)
	at org.jooq.impl.SelectImpl.fetchOne(SelectImpl.java:3074)
	at org.jooq.impl.DefaultDSLContext.fetchOne(DefaultDSLContext.java:4535)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4573)
	at org.jooq.impl.DefaultDSLContext.fetchValue(DefaultDSLContext.java:4587)
	at org.jooq.test.all.testcases.ConverterTests.lambda$5(ConverterTests.java:262)
	at org.jooq.test.jOOQAbstractTest.bindAndInline(jOOQAbstractTest.java:1127)
	at org.jooq.test.BaseTest.bindAndInline(BaseTest.java:1628)
	at org.jooq.test.all.testcases.ConverterTests.testConvertOnlyOnce(ConverterTests.java:240)
	at org.jooq.test.jOOQAbstractTest.testConvertOnlyOnce(jOOQAbstractTest.java:5422)
```

We should certainly be able to cache the converted value between the two passes - the information should still be available. Will look into this.

**Comment by lukaseder** [comment]
There are a few reasons why the converted value needs to be available when generating the SQL string:

- It could be `null`, in case of which the generated SQL might be different in some dialects
- Generated casts might depend on the value (e.g. precision, scale)
- The actual `<T>` type might be relevant as it might differ from the `Class<T>` type returned from `Converter::toType` (specifically, when using `SQLDataType.OTHER`)
- The bind value is inlined (in case of which there will be only one call, as the bind stack trace is omitted)

If the bind value is not inlined, then there's the second stack trace from above, which needs to re-convert the `U` value to `T`, prior to sending it to the JDBC driver.

**Comment by lukaseder** [comment]
It's not trivial to find a performant and clean solution to lazily converting the bind value only once between the two passes. We could replace the `Param` by a new one that references the `T` type, not the `U` type, but a custom `Binding` would still need to be maintained, and we don't know if the custom `Binding` can work with the `T` type.

I'll put this issue on hold for now until there's an obvious solution.

A workaround on your side could be to use some `ThreadLocal` (or otherwise global) `HashMap` where you temporarily store the converted data, and prevent a second conversion in your `EncryptionConverter`. In an `ExecuteListener`, you can then clean up this global map on the `ExecuteListener::executeEnd` event.

**Comment by lukaseder** [comment]
Looking at this again. There's a memory / CPU tradeoff to be made here. If we want to cache the result of the first conversion (from the SQL string generation), we need to store it *somewhere*. The CPU side is only relevant to those users who actually use converters, and even then, only rarely, because most converters are quite fast. Compared to the SQL string generation, we don't save any significant CPU when caching these values, leaving only the few edge cases where conversion is quite heavy as in your case.

All of this is not in favour of a caching solution that is enabled by default. This doesn't even touch on the risk of introducing a regression.

I think the thread-local approach is optimal here. Create some:

```java
final ThreadLocal<Entry<String, String>> tl = ...
```

Where the key is the input value, the value is the output value, and the threadlocal always contains the thread's last conversion pair. Here's what I mean:

```java
abstract class CachingConverter<T, U> extends AbstractConverter<T, U> {

    final ThreadLocal<SimpleEntry<T, U>> tl = new ThreadLocal<>();

    public CachingConverter(Class<T> fromType, Class<U> toType) {
        super(fromType, toType);
    }

    @Override
    public final U from(T t) {
        U result;
        SimpleEntry<T, U> e = tl.get();

        if (e == null)
            tl.set(e = new SimpleEntry<>(t, result = from0(t)));
        else if (Objects.equals(t, e.getKey()))
            result = e.getValue();
        else
            e.setValue(result = from0(t));

        return result;
    }

    abstract U from0(T t);
}
```

Optionally, this can be done for the `to()` method as well, though there aren't 2 calls to it, so it's not necessary in your case. Of course, there's a security risk here if you're sensitive to this kind of problem. The values being stuck in memory and possibly "shared" between sessions might not be acceptable, so maybe, the thread local needs to be cleaned up in an `ExecuteListener`, or even stored elsewhere.

I'll give this a bit more thought and then possibly close this as won't fix. I can't think of an acceptable solution to this issue. If one pops up later on, we can still fix it.

**Comment by lukaseder** [comment]
Well, to prevent traversing the entire `QueryPart` tree, we collect all bind values into an internal `QueryPartList<Param<?>>` while rendering the SQL, so at least wo don't necessarily require any additional `List` to collect the converted values. Remains to be seen if we can store the converted values in that list, instead of the original ones, or at least cache both values there.

**Comment by lukaseder** [comment]
Even then, it's quite the refactoring... Will revisit again in the future.

**Comment by lukaseder** [comment]
I just had a funny case of this where I used `StringBuilder` in a `Converter<StringBuilder, StringBuilder>` test. The converter mutated the user `StringBuilder` value (which it obviously shouldn't, it should produce a new instance. But the fact that the converter was called twice produced wrong behaviour in this case.
